### PR TITLE
Improve database sorting for user profiles

### LINKED ISSUE #1624: Handle user profile sorting with incomplete names
Sorting user profiles by first and last name currently fails when not both are specified. This happens for special users such as, e.g., "Student Representatives" who do not have a first and last name specified.

For example, when opening a semester's grade reminder page, loading the page currently fails when at least one of the responsible users of a course listed on the page does not have a first and last name specified (`AttributeError at /staff/semester/<id>/grade_reminder 'NoneType' object has no attribute 'lower'`).

As a solution, if only a last name is specified, the first name can be ignored. If both values are None, the email must be used for sorting.

These places need to be checked and updated:
- `evaluation.models.responsibles_names`
- `staff.views.semester_preparation_reminder`
- `staff.views.semester_grade_reminder`

### PR DESCRIPTION
**Pull request by postmartem**
Closes #1624

### CODE REVIEW

**Comment by niklasmohrin** [comment]
The `FieldException` is a bug in django, they fixed it 3 months ago. We will update in the next couple of days

File: evap/evaluation/models.py [hunk 1]
```diff
@@ -356,10 +356,7 @@ def midterm_grade_documents(self):
 
     @cached_property
     def responsibles_names(self):
-        ordered_responsibles = sorted(
-            self.responsibles.all(), key=lambda responsible: (responsible.last_name, responsible.full_name)
-        )
-        return ", ".join([responsible.full_name for responsible in ordered_responsibles])
+        return ", ".join([responsible.full_name for responsible in self.responsibles.all()])
```
Comment (niklasmohrin) [inline evap/evaluation/models.py:359]:
```suggestion
        return ", ".join(responsible.full_name for responsible in self.responsibles.all())
```

File: evap/evaluation/models.py [hunk 2]
```diff
@@ -1547,7 +1544,7 @@ class UserProfile(AbstractBaseUser, PermissionsMixin):
     is_active = models.BooleanField(default=True, verbose_name=_("active"))
 
     class Meta:
-        ordering = ["last_name", "first_name", "email"]
+        ordering = [Lower(Concat("last_name", "first_name", "email"))]
```
Comment (niklasmohrin) [inline evap/evaluation/models.py:1547]: sorry, we only now noticed that we may not want `Concat` afterall; it should just sort for the `Lower` of each individually

File: evap/evaluation/tests/test_models.py [hunk 3]
```diff
@@ -670,6 +670,19 @@ def test_get_sorted_due_evaluations(self):
         sorted_evaluations = student.get_sorted_due_evaluations()
         self.assertEqual(sorted_evaluations, [(evaluations[1], 0), (evaluations[0], 0), (evaluations[2], 1)])
 
+    def test_can_be_sorted(self):
```
Comment (niklasmohrin) [inline evap/evaluation/tests/test_models.py:673]: I mean, they _can_ be sorted, but _are_ they? And is that sorting correct? Maybe `test_correct_sorting`?

File: evap/evaluation/tests/test_models.py [hunk 4]
```diff
@@ -670,6 +670,19 @@ def test_get_sorted_due_evaluations(self):
...
+        baker.make(
+            UserProfile,
+            last_name=iter(["B", "A", None]),
+            first_name=iter(["X", None, None]),
+            email=iter(["ax@example.com", "b@example.com", "unnamed@example.com"]),
+            _quantity=3,
+            _bulk_create=True,
+        )
+        user_list = list(UserProfile.objects.all())
+        email_list = [user.email for user in user_list]
```
Comment (niklasmohrin) [inline evap/evaluation/tests/test_models.py:683]:
```suggestion
        email_list = [user.email for user in UserProfile.objects.all()]
```

**Review by niklasmohrin — CHANGES_REQUESTED** [review]

File: evap/evaluation/migrations/0131_alter_userprofile_options.py [hunk 5]
```diff
@@ -0,0 +1,18 @@
+# Generated by Django 4.0.6 on 2022-08-15 20:12
```
Comment (niklasmohrin) [inline evap/evaluation/migrations/0131_alter_userprofile_options.py:1]: one more thing: Can you rename this file to `0131_userprofile_ordering.py`?

File: evap/evaluation/tests/test_models.py [hunk 6]
```diff
@@ -670,6 +670,18 @@ def test_get_sorted_due_evaluations(self):
         sorted_evaluations = student.get_sorted_due_evaluations()
         self.assertEqual(sorted_evaluations, [(evaluations[1], 0), (evaluations[0], 0), (evaluations[2], 1)])
 
+    def test_correct_sorting(self):
+        baker.make(
+            UserProfile,
+            last_name=iter(["B", "A", None]),
```
Comment (niklasmohrin) [inline evap/evaluation/tests/test_models.py:676]: Strictly, this doesn't test that a) we `Lower` the fields and b) that there are more sorting keys than just the last name

**Review by He3lixxx — APPROVED** [review]

**Review by niklasmohrin — APPROVED** [review]

**Pull request merged**